config/types/files.go

```diff
@@ -129,6 +130,19 @@ func (fc FileContents) Validate() report.Report {
 	return report.Report{}
 }
 
+func makeDataUrl(contents []byte) string {
+	opaque := "," + dataurl.Escape(contents)
+	b64 := ";base64," + base64.StdEncoding.EncodeToString(contents)
+	if len(b64) < len(opaque) {
+		opaque = b64
+	}
+	uri := (&url.URL{
+		Scheme: "data",
+		Opaque: opaque,
+	}).String()
+	return uri
+}
```

Comment:
Can we have some tests for this code? I'm curious under what conditions base64 will actually be smaller.

I think we should also test that Ignition is able to parse the produced output somehow?

Reply:
Let me think about tests. The only way to test that ignition can parse it is by checking that the dataurl library is able to parse it.

The output is smaller when the input contains many non-ascii characters, an example would be a gzipped file. So if you do:
```
$ echo hello world | gzip -c - | base64
H4sIAAAAAAAAA8tIzcnJVyjPL8pJ4QIALTsIrwwAAAA=
```
and then:
```
storage:
  files:
  - path: /hello.gz
    mode: 0644
    contents:
      inline: !!binary |
        H4sIAAAAAAAAA8tIzcnJVyjPL8pJ4QIALTsIrwwAAAA=
```
This ends up like this after this commit:
`{"ignition":{"config":{},"security":{"tls":{}},"timeouts":{},"version":"2.3.0"},"networkd":{},"passwd":{},"storage":{"files":[{"filesystem":"root","path":"/hello.gz","contents":{"source":"data:;base64,H4sIAAAAAAAAA8tIzcnJVyjPL8pJ4QIALTsIrwwAAAA=","verification":{}},"mode":420}]},"systemd":{}}`
and before it was:
`{"ignition":{"config":{},"security":{"tls":{}},"timeouts":{},"version":"2.3.0"},"networkd":{},"passwd":{},"storage":{"files":[{"filesystem":"root","path":"/hello.gz","contents":{"source":"data:,%1F%8B%08%00%00%00%00%00%00%03%CBH%CD%C9%C9W(%CF%2F%CAI%E1%02%00-%3B%08%AF%0C%00%00%00","verification":{}},"mode":420}]},"systemd":{}}`

Reply:
I think we could have a test testing both variants, base64 and "plain", both validated using some ignition validator.

Reply:
I tried adding a test in Ignition, but it looks like the v3 update will help because it includes a GitHub Action to run the tests and maybe some fixes to execute them correctly: https://github.com/flatcar-linux/ignition/pull/30

For now I think it's fine if this here is tested manually since we also have kola tests in mantle that use base64 and will notice if Ignition would have broken base64 support.

Reply:
The Ignition test is merged downstream and upstream

Reply:
I'd rather have tests locally, but well...